system/src/system_connection_manager.cpp

```diff
@@ -96,7 +96,13 @@ ConnectionManager* ConnectionManager::instance() {
 void ConnectionManager::setPreferredNetwork(network_handle_t network, bool preferred) {
     if (preferred) {
         if (network != NETWORK_INTERFACE_ALL) {
-            preferredNetwork_ = network;    
+            preferredNetwork_ = network;
+
+            // If preferred network is set, and it is up, move cloud connection to it immediately
+            if (network_ready(spark::Network.from(preferredNetwork_), 0, nullptr)) {
+                auto options = CloudDisconnectOptions().graceful(true).reconnect(true).toSystemOptions();
```

Comment:
Will `graceful()` close the session?  And don't we want the session to continue?

Reply:
Yes. `graceful()` closes the cloud connection. A graceful disconnection means we will [send a message](https://github.com/particle-iot/device-os/blob/acm-failover-speed/system/src/system_task.cpp#L671-L680) to the device service notifying that we plan on disconnecting. I believe this is the preferred way to disconnect from the device side if possible. 

The new `reconnect()` option means we want the system to immediately re-establish the cloud connection/socket as soon as possible. This will continue the session on the new network interface

Reply:
So the session is resumed without all of the handshaking overhead?

Reply:
Yes, here is a log of switching a connection when `Ethernet.prefer()` is called
```
0006965201 [app] INFO: Prefer Ethernet
0006965290 [system] INFO: Cloud: disconnecting
0006965296 [comm.coap] TRACE: Sending CoAP message
0006965304 [comm.coap] TRACE: CON POST /x size=9 token= id=25
0006965314 [comm] INFO: Waiting for Confirmed messages to be sent.
0006965847 [comm.coap] TRACE: Received CoAP message
0006965852 [comm.coap] TRACE: ACK 0.00  size=4 token= id=25
0006966315 [comm] INFO: All Confirmed messages sent: client(yes) server(yes)
0006966325 [system.ledger] TRACE: Disconnected
0006966331 [system] INFO: Cloud: disconnected
0006966337 [system] INFO: Cloud: connecting
0006966343 [system] INFO: Read Server Address = type:1,domain:$id.udp.particle.io
0006966354 [system] INFO: Loaded cloud server address and port from session data
0006966366 [system] TRACE: Address type: 1
0006966371 [system] TRACE: Cloud socket=0, family=2, type=2, protocol=17
0006966379 [system] INFO: Cloud socket=0, connecting to 18.214.12.187#5684
0006966393 [system] INFO: Bound cloud socket to lwip interface en2
0006966403 [system] TRACE: Cloud socket=0, connected to 18.214.12.187#5684
0006966414 [system] TRACE: Updating cloud keepalive for AF_INET: 1380000 -> 25000
0006966426 [system] TRACE: Applying new keepalive interval now
0006966436 [system] INFO: Cloud socket connected
0006966442 [system] INFO: Starting handshake: presense_announce=0
0006966451 [comm.protocol.handshake] INFO: Establish secure connection
0006966462 [comm.dtls] INFO: session has 0 uses
0006966472 [comm.dtls] INFO: (CMPL,RENEG,NO_SESS,ERR) restoreStatus=0
0006966482 [comm.dtls] INFO: out_ctr 0,1,0,0,0,0,0,31, next_coap_id=19
0006966492 [comm.dtls] INFO: restored session from persisted session data. next_msg_id=25
0006966502 [comm.dtls] INFO: session cmd (CLS,DIS,MOV,LOD,SAV): 2
0006967417 [ncp.at] TRACE: > AT+CFUN?
0006967426 [ncp.at] TRACE: < +CFUN: 1
0006967428 [ncp.at] TRACE: < OK
0006967430 [ncp.at] TRACE: > AT+CCID
0006967433 [ncp.at] TRACE: < +CCID: 89883070000037323357
0006967435 [ncp.at] TRACE: < OK
0006967437 [ncp.at] TRACE: > AT+CGSN
0006967440 [ncp.at] TRACE: < 862044067617865
0006967443 [ncp.at] TRACE: < OK
0006967445 [ncp.at] TRACE: > AT+CGMR
0006967447 [ncp.at] TRACE: < BG95M5LAR02A03
0006967449 [ncp.at] TRACE: < OK
0006967452 [comm.protocol.handshake] INFO: Skipping HELLO message
0006967460 [comm.coap] TRACE: Sending CoAP message
0006967469 [comm.coap] TRACE: CON 0.00  size=4 token= id=26
0006967483 [system] INFO: cloud connected from existing session.
0006967490 [comm.coap] TRACE: Sending CoAP message
0006967497 [comm.coap] TRACE: CON POST /E/particle/device/updates/enabled size=44 token= id=27
0006967510 [comm.coap] TRACE: Sending CoAP message
0006967517 [comm.coap] TRACE: CON POST /E/particle/device/updates/forced size=44 token= id=28
0006967530 [system] INFO: Sending application DESCRIBE
0006967539 [comm.protocol] INFO: Checksum has not changed; not sending application DESCRIBE
0006967555 [system] INFO: Sending subscriptions
0006967561 [comm.protocol] INFO: Checksum has not changed; not sending subscriptions
0006967571 [system] TRACE: Waiting until all handshake messages are processed by the protocol layer
0006967587 [comm.dtls] INFO: session cmd (CLS,DIS,MOV,LOD,SAV): 4
0006967599 [comm.coap] TRACE: Received CoAP message
0006967604 [comm.coap] TRACE: ACK 0.00  size=4 token= id=26
0006967613 [comm.coap] TRACE: Received CoAP message
0006967622 [comm.coap] TRACE: ACK 0.00  size=4 token= id=27
0006967633 [comm.coap] TRACE: Received CoAP message
0006967642 [comm.coap] TRACE: ACK 0.00  size=4 token= id=28
0006967649 [system] INFO: All handshake messages have been processed
0006967757 [system] INFO: Cloud connected
```

Total time to switch is ~2.5 seconds, session is resumed

Reply:
Thank you for the explanation.